spyderlib/widgets/sourcecode/base.py

```diff
@@ -811,12 +811,17 @@ def __move_line_or_selection(self, after_current_line=True):
             start_pos = cursor.position()
             cursor.movePosition(QTextCursor.NextBlock)
             end_pos = cursor.position()
+            # check if on last line
+            if end_pos == start_pos:
+                cursor.movePosition(QTextCursor.End)
+                cursor.insertText(os.linesep)

```

Comment:
This line introduces a **new** empty line in the file, which I think it's not good. For example, if these are the last two lines of a file (i.e. without a new line at the end):

```python
f(1, 2)
g(1, 2)
```

we have:

1. Pressing `Alt+Down` on `g` introduces an empty line after it.
2. Pressing `Alt+Up` on `g` interchanges with `f`, but creates an empty line below it.
3. Pressing `Alt+Down` on `f` doesn't interchange it with `g`.